### PR TITLE
Add backwards compatibility policy

### LINKED ISSUE #595: Do not break public interfaces, have a policy
### Context

We need to have a backward compatibility policy.

This policy needs to specify what our public interfaces are. Example interfaces:
- conda-store is used as a library via Python
- conda-store web API is used via HTTP.

This policy needs to document how we make changes to those interfaces. Examples:
- a warning is raised for N conda-store releases before removing functionality
- an error is raised after that.

As an example, here's BC/FC policy for PyTorch: https://github.com/pytorch/pytorch/wiki/PyTorch's-Python-Frontend-Backward-and-Forward-Compatibility-Policy

An example where a Python API change broke nebari: https://github.com/conda-incubator/conda-store/issues/593#issuecomment-1732244137.

### Value and/or benefit

We don't break other software that depends on conda-store, like nebari.
When we introduce changes, things are easier to debug because an error/warning is raised.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by dcmcand**
Closes #595 
This pull request:
Creates a Backwards Compatibility Policy for the conda-store project as worked out by @nkaretnikov, @costrouc, and myself.

@costrouc and @nkaretnikov , please ensure that I have accurately captured our discussion.

@trallard I welcome any comments. Also, where in the repo would you like this policy to live? I was unclear on that so put it in the root for the meantime.

### CODE REVIEW

**Comment by pavithraes** [comment]
Thanks for putting this together.

> Also, where in the repo would you like this policy to live? I was unclear on that so put it in the root for the meantime.

This needs to live at `docusaurus-docs/community/policies/backwards-compatibility.md`. Note that the `policies` directory doesn't exist on the `main` branch, so we'll need to create it in this PR. Alternatively, I'm adding a placeholder for this policy doc in #685, if we merge that first, you can move this doc to the placeholder page directly.

**Review by nkaretnikov — CHANGES_REQUESTED** [review]
@dcmcand Good overall! Left some comments, looking forward to the final version!

See what Pavithra wrote regarding where this file should be placed.

File: BackwardsCompatibilityPolicy.md
Comment (nkaretnikov) [reply BackwardsCompatibilityPolicy.md:17]: Nit: use the `-` list format instead where the order doesn't matter

**Review by trallard — CHANGES_REQUESTED** [review]
Thanks @dcmcand - I went through and added some comments and suggestions. 
I do not why but the API sections feel a lot more solid than the ones pertaining Python code - the latter seems like they get a bit too into the weeds (methods, classes, values) and is thin on the more procedural/implementation side.

I like for example NumPy's policy https://numpy.org/neps/nep-0023-backwards-compatibility.html#implementing-deprecations-and-removals 

1. It makes clear when a deprecation is needed vs being too prescriptive or granulas
2. provides clear steps needed to implement a deprecation (add deprecation warning, add to the notes, add to the docs) for example within this PR

> Deprecated classes, methods, and functions should have a comment and a warning (if possible) in the code stating why they are deprecated and what to use instead. This will encourage developers not to use them without breaking existing code.....

Perhaps this should be tightened to add: deprecation warning, include the version number of the release where this was deprecated, provide alternatives (if applicable), add to release notes and document accordingly with visible deprecation status.

3. Also includes process on how to go about deprecations -  se https://numpy.org/neps/nep-0023-backwards-compatibility.html#decision-making - this is missing.  We should have this process to ensure deprecations do not unnoticed but are intentional instead and approved/agreed on by the core team. We can think of a process like

   1.  Open an RFC issue in the project repository 
   2.  Follow a consensus-based decision making approach (ref: https://producingoss.com/en/producingoss.html#consensus-democracy https://www.seedsforchange.org.uk/shortconsensus) 
   3. If consensus is reached move to implementation and follow the rest of the guidelines described in this doc

**Review by nkaretnikov — CHANGES_REQUESTED** [review]
@dcmcand The current version looks great, thanks!

I only have 1 or 2 concrete suggestions, for wording and spelling, the rest of my comments are mostly stylistic nitpicks, which I think will make the source doc easier to read.

I agree with most of Tania's suggestions, too, and clarified where I disagree or have questions.

I'd also prefer we wrap this to 80 chars to match our source code (we use `line-length = 88` for Python code), but I don't feel strongly about it if it's too much work.

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (trallard) [reply docusaurus-docs/community/policies/backwards-compatibility.md:20]: I actually prefer no points in short bullets so will leave it to Chuck's preference

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (dcmcand) [reply docusaurus-docs/community/policies/backwards-compatibility.md:13]: not necessarily. If an api endpoint formerly had an optional field but we make it  a required field, that is adding a new requirement. Or if a public python function previously had a default value for a parameter and we remove that default, that is adding a requirement. It means we are adding something required that is either new or was previously optional.

Comment (dcmcand) [reply docusaurus-docs/community/policies/backwards-compatibility.md:13]: I have added language to clarify

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (dcmcand) [reply docusaurus-docs/community/policies/backwards-compatibility.md:103]: yeah, I think this should be backwards.

**Comment by dcmcand** [comment]
> Thanks, @dcmcand - I went through and added some comments and suggestions. I do not know why, but the API sections feel a lot more solid than the ones about Python code - the latter seems like they get too into the weeds (methods, classes, values) and is thin on the more procedural/implementation side.
> 
> I like, for example, NumPy's policy https://numpy.org/neps/nep-0023-backwards-compatibility.html#implementing-deprecations-and-removals
> 
>     1. It makes clear when a deprecation is needed vs being too prescriptive or granular
> 
>     2. It provides clear steps needed to implement a deprecation (add deprecation warning, add to the notes, add to the docs) for example within this PR
> 
> 
> > Deprecated classes, methods, and functions should have a comment and a warning (if possible) in the code stating why they are deprecated and what to use instead. This will encourage developers not to use them without breaking existing code.....
> 
> Perhaps this should be tightened to add a deprecation warning, include the version number of the release where this was deprecated, provide alternatives (if applicable), add to release notes and document accordingly with visible deprecation status. (you get the gist)
> 
>     3. It Also includes the process on how to go about deprecations -  see: https://numpy.org/neps/nep-0023-backwards-compatibility.html#decision-making - this needs to be included.  We should have this process to ensure deprecations do not go unnoticed but are intentional instead and approved/agreed on by the core team. We can think of a process like
>        
>        1. Open an RFC issue in the project repository
>        2. Follow a consensus-based decision-making approach (ref: https://producingoss.com/en/producingoss.html#consensus-democracy https://www.seedsforchange.org.uk/shortconsensus)
>        3. If consensus is reached, move to implementation and follow the rest of the guidelines described in this doc

This still needs added @trallard but please feel free to take a look at everything else.

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (dcmcand) [reply docusaurus-docs/community/policies/backwards-compatibility.md:243]: Take a look at what I added and see if you think it is better. If you still thinks it needs work, I am happy to add more or rework what is there.

**Review by nkaretnikov — CHANGES_REQUESTED** [review]
@dcmcand Thanks for working on this! Unfortunately, I've found some issues caused by a recent commit that wrapped long lines. Also, the lint issues need to be fixed to make the tests pass. Once all these issues are addressed, I'll be happy to accept the PR.

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (trallard) [inline docusaurus-docs/community/policies/backwards-compatibility.md:51]:
```suggestion
ones.
- Columns and tables should not be renamed. Aliases should be used for
```

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (nkaretnikov) [inline docusaurus-docs/community/policies/backwards-compatibility.md:32]:
```suggestion
:::info[TL;DR]
```

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (nkaretnikov) [inline docusaurus-docs/community/policies/backwards-compatibility.md:45]:
```suggestion
changes. Databases hold the state for the application. Introducing breaking
```

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (nkaretnikov) [inline docusaurus-docs/community/policies/backwards-compatibility.md:47]:
```suggestion
earlier versions of conda-store. To maintain backwards compatibility we follow
```

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (nkaretnikov) [inline docusaurus-docs/community/policies/backwards-compatibility.md:63]:
```suggestion
- adding a parameter to the return value of an endpoint
- making a previously mandatory input optional.
```

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (nkaretnikov) [inline docusaurus-docs/community/policies/backwards-compatibility.md:70]:
```suggestion
- removing a parameter from the return value
- altering the meaning of a value
- making a formerly optional parameter mandatory
```

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (nkaretnikov) [inline docusaurus-docs/community/policies/backwards-compatibility.md:88]:
```suggestion
`example.com/api/experimental/user` route. This allows conda-store contributors
```

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (nkaretnikov) [inline docusaurus-docs/community/policies/backwards-compatibility.md:90]:
```suggestion
a new version of an API endpoint. Using the `experimental` namespace is not
```

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (nkaretnikov) [inline docusaurus-docs/community/policies/backwards-compatibility.md:108]:
````suggestion
```
https://example.com/api/v1/user
https://example.com/api/v2/user
```
````

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (nkaretnikov) [inline docusaurus-docs/community/policies/backwards-compatibility.md:113]:
```suggestion
Explanation: This route has breaking changes between `v1` and `v2`. Code
written for the `v1` endpoint will continue to function, but all new features
will only be available in the `v2` version of the endpoint. This empowers users
to upgrade when they wish to rather than forcing them to do so.
```

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (nkaretnikov) [inline docusaurus-docs/community/policies/backwards-compatibility.md:118]:
````suggestion
```
https://example.com/api/v1/user
```
````

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (nkaretnikov) [inline docusaurus-docs/community/policies/backwards-compatibility.md:125]:
````suggestion
```
https://example.com/api/experimental/user
```
````

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (nkaretnikov) [inline docusaurus-docs/community/policies/backwards-compatibility.md:168]:
```suggestion
- version number of the release where this was deprecated
- provide suggestions for alternatives (if possible)
- provide justification for the removal (such as a link to the issue
  or CVE that necessitated the removal).
```

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (nkaretnikov) [inline docusaurus-docs/community/policies/backwards-compatibility.md:178]:
```suggestion
Private examples:
- `_private_func_or_method`
- `_PrivateClass`
- `_PRIVATE_VAR`.
```

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (nkaretnikov) [inline docusaurus-docs/community/policies/backwards-compatibility.md:180]:
```suggestion
Public examples:
- `public_func_or_method`
- `PublicClass`
- `PUBLIC_VAR`.
```

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (nkaretnikov) [inline docusaurus-docs/community/policies/backwards-compatibility.md:230]:
```suggestion
notes and documentation. This should include:
```

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (nkaretnikov) [inline docusaurus-docs/community/policies/backwards-compatibility.md:234]:
```suggestion
- version number of the release where this was deprecated
- provide suggestions for alternatives (if possible)
- provide a reason for the deprecation or removal (such
  as a link to a CVE or issue that necessitated the removal).
```

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (nkaretnikov) [inline docusaurus-docs/community/policies/backwards-compatibility.md:251]:
```suggestion
In a public class, a breaking change is one that changes or removes attributes
or methods or alters their meanings. Rather than changing methods or
attributes, new methods or attributes should be added if needed. For example,
if you wanted to change a user id from an `int` to a `uuid`, a new attribute
`User.uuid` should be added, and all new code should use `User.uuid`. Existing
methods can use the new attribute or methods as well as long as that doesn't
introduce breaking changes for the method.
```

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (nkaretnikov) [inline docusaurus-docs/community/policies/backwards-compatibility.md:269]:
```suggestion
it should be asynchronous, a new function called `list_envs_async` should be
```

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (nkaretnikov) [inline docusaurus-docs/community/policies/backwards-compatibility.md:52]:
```suggestion
- New columns or tables should be added instead of removing or altering existing
  ones. 
- Columns and tables should not be renamed. Aliases should be used for
  poorly named existing columns or tables.
```

**Review by nkaretnikov — CHANGES_REQUESTED** [review]
@dcmcand I've updated my comments using the "suggestions" feature. PTAL

**Comment by nkaretnikov** [comment]
@dcmcand The linter is still failing, will review once the CI is green.

**Review by nkaretnikov — APPROVED** [review]
@dcmcand LGTM, thank you! I've looked at the Docusaurus-rendered version locally to make sure it's displayed properly, too.

@trallard Do you have any additional comments? If not, I think this is ready for merge. We can always address minor details in followups.

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (trallard) [inline docusaurus-docs/community/policies/backwards-compatibility.md:115]:
```suggestion
**Explanation**: This route has breaking changes between `v1` and `v2`. Code
```

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (trallard) [inline docusaurus-docs/community/policies/backwards-compatibility.md:127]:
```suggestion
**Explanation**: This route has never had breaking changes introduced. Any code
```

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (trallard) [inline docusaurus-docs/community/policies/backwards-compatibility.md:136]:
```suggestion
**Explanation**: This is an experimental route. It can be changed or removed at any
```

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (trallard) [inline docusaurus-docs/community/policies/backwards-compatibility.md:217]:
```suggestion
:::important
Tests are never considered to be part of the public API. Any code within the
`tests/` directory is always considered to be private.
:::

```

File: docusaurus-docs/community/policies/backwards-compatibility.md
Comment (trallard) [reply docusaurus-docs/community/policies/backwards-compatibility.md:257]: This looks better, let's merge and if needed iterate on this

**Review by trallard — APPROVED** [review]
Thanks @dcmcand all looks pretty good now. 
I left a couple of nits (formatting) that you can or not include, but otherwise this should be ready for merging

**Pull request merged**
